server/console_authenticate.go

```diff
@@ -81,10 +86,20 @@ func (s *ConsoleServer) Authenticate(ctx context.Context, in *console.Authentica
 			role = console.UserRole_USER_ROLE_ADMIN
 			uname = in.Username
 			id = uuid.Nil
+		} else {
+			if lockout, until := s.loginAttemptCache.Add(s.config.GetConsole().Username, ip); lockout != LockoutTypeNone {
```

Comment:
@zyro , @ftkg ,

there is a bug in `extractClientAddressFromContext` allowing  IP address to  be spoofed, making it easy to bypass this limiter.  Bug is that IP address extracted as a first element of x-forwarded-for header, not rightmost (N - 1), where N is number of known proxies, as a result IP address as nakama knows it is entirely controlled by user. 

Even without IP spoofing, with enough IP addresses available it looks like it is quite possible to exhaust server memory  by authenticating with sufficiently log usernames.  If we truncate username to something like 20 chars before storing in the cache it will be much harder to pull it off

Reply:
Per [this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For#syntax) the left-most address is supposed to indicate the client address, assuming trustworthy proxies across the chain. Either way we'll leave IP-based restrictions for a later stage.

Account lockouts are only tracked for valid accounts that actually exist, a bad actor cannot fill up the list with arbitrary usernames.